File: client_E_R.java
import * as socket from 'socket';
import * as sys from 'sys';
import * as threading from 'threading';
var connexion, host, port, th_E, th_R;
host = "127.0.0.1";
port = 40000;

class ThreadReception extends threading.Thread {
  /* objet thread gérant la réception des messages */
  constructor(conn) {
    threading.Thread.__init__(this);

    this.connexion = conn;
  }

  run() {
    var message_recu;

    while (1) {
      message_recu = this.connexion.recv(1024).decode();
      console.log("*" + message_recu + "*");

      if (message_recu === "" || message_recu.upper() === "FIN") {
        break;
      }
    }

    th_E._Thread__stop();

    console.log("Client arr\u00eat\u00e9. Connexion interrompue.");
    this.connexion.close();
  }

}

class ThreadEmission extends threading.Thread {
  /* objet thread gérant l'émission des messages */
  constructor(conn) {
    threading.Thread.__init__(this);

    this.connexion = conn;
  }

  run() {
    var message_emis;

    while (1) {
      message_emis = input();
      this.connexion.send(message_emis.encode());
    }
  }

}

connexion = socket.socket(socket.AF_INET, socket.SOCK_STREAM);

try {
  connexion.connect([host, port]);
} catch (e) {
  if (e instanceof socket.error) {
    console.log("La connexion a \u00e9chou\u00e9.");
    sys.exit();
  } else {
    throw e;
  }
}

console.log("Connexion \u00e9tablie avec le serveur.");
th_E = new ThreadEmission(connexion);
th_R = new ThreadReception(connexion);
th_E.start();
th_R.start();

File: client_E_R.py
# Définition d'un client réseau gérant en parallèle l'émission
# et la réception des messages (utilisation de 2 THREADS).

host = '127.0.0.1'
port = 40000

import socket, sys, threading

class ThreadReception(threading.Thread):
    """objet thread gérant la réception des messages"""
    def __init__(self, conn):
        threading.Thread.__init__(self)
        self.connexion = conn           # réf. du socket de connexion
        
    def run(self):
        while 1:
            message_recu = self.connexion.recv(1024).decode()
            print("*" + message_recu + "*")
            if message_recu =='' or message_recu.upper() == "FIN":
                break
        # Le thread <réception> se termine ici.
        # On force la fermeture du thread <émission> :
        th_E._Thread__stop()
        print("Client arrêté. Connexion interrompue.")
        self.connexion.close()
    
class ThreadEmission(threading.Thread):
    """objet thread gérant l'émission des messages"""
    def __init__(self, conn):
        threading.Thread.__init__(self)
        self.connexion = conn           # réf. du socket de connexion
        
    def run(self):
        while 1:
            message_emis = input()
            self.connexion.send(message_emis.encode())

# Programme principal - Établissement de la connexion :
connexion = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
try:
    connexion.connect((host, port))
except socket.error:
    print("La connexion a échoué.")
    sys.exit()    
print("Connexion établie avec le serveur.")
            
# Dialogue avec le serveur : on lance deux threads pour gérer
# indépendamment l'émission et la réception des messages :
th_E = ThreadEmission(connexion)
th_R = ThreadReception(connexion)
th_E.start()
th_R.start()